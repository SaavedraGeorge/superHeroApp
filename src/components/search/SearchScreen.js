import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesbyName } from '../../selector/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import queryString from 'query-string';



export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  
  const {q = ''} = queryString.parse(location.search);

  const [values, handleInputChange] = useForm({
    searchText: q,
  });

  const {searchText} = values;


  
  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  }
 
   const heroesFiltered =  useMemo(() => getHeroesbyName(q), [q]);

 
  
  return(
      <>
          <h1>Search Screen</h1>
          <hr />
          <div className="row">
            <div className="col-5">
              <h4>Búsqueda de personajes</h4>
              <hr />
              <form onSubmit={handleSearch}>
                <input 
                  type="text" 
                  placeholder="Búsqueda de personajes..."
                  className="form-control"
                  name="searchText"
                  value={searchText}
                  onChange={handleInputChange}
                />
                <button type="submit" className="btn btn-outline-primary mt-3 col-12">
                  Buscar
                </button>
              </form>
            </div>
            <div className="col-7">
              <h4>Resultados Búsqueda</h4>
              <hr />
              {
                (q === '') 
                  ? <div className="alert alert-info animate__animated animate__fadeIn">Busca un heroe</div>
                  : ( heroesFiltered.length === 0)
                    && <div className="alert alert-danger animate__animated animate__fadeIn">No hay resultados para la busqueda: <b>{q}</b> </div>
              }
              {
                heroesFiltered.map(heroe => (
                  <HeroCard key={heroe.id}
                  {...heroe} />
                ))
              }
            </div>
          </div>
      </>
  );
};
