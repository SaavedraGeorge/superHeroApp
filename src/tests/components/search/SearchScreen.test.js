import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";


const mockNavigate = jest.fn();

jest.mock('react-router-dom', ()=> ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('puevas en <searchScreen/>', () => {
  test('should show correctly', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe');
  });
  test('should show correctly batman Search', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman']}>
            <SearchScreen />
        </MemoryRouter>
    )
    expect(wrapper.find('input').prop('value')).toBe('batman');
  });
  test('should show error in the search', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=wrong']}>
          <SearchScreen />
      </MemoryRouter>
  )
  expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados para la busqueda: wrong');
  });
  test('debe de llamar el navigate a la nueva pantalla ', () => {
        
    const wrapper = mount( 
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen /> 
        </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
        target: {
            name: 'searchText',
            value: 'batman'
        }
    });

    wrapper.find('form').prop('onSubmit')({
        preventDefault(){}
    })

    expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');


})
  
});
