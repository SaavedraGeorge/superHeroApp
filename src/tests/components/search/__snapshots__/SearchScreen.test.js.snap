// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`puevas en <searchScreen/> should show correctly 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Búsqueda de personajes
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Búsqueda de personajes..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3 col-12"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados Búsqueda
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
        Busca un heroe
      </div>
    </div>
  </div>,
]
`;
